src/subdevice/mod.rs

```diff
@@ -771,6 +771,111 @@ where
         Ok(())
     }
 
+    #[cfg(feature = "std")]
+    /// Read all sub-indices of a SDO array object.
+    ///
+    /// Example:
+    /// ```rust,no_run
+    /// # use ethercrab::{
+    /// #     error::Error, MainDevice, MainDeviceConfig, PduStorage, Timeouts, std::ethercat_now
+    /// # };
+    /// # static PDU_STORAGE: PduStorage<8, 32> = PduStorage::new();
+    /// # let (_tx, _rx, pdu_loop) = PDU_STORAGE.try_split().expect("can only split once");
+    /// # let maindevice = MainDevice::new(pdu_loop, Timeouts::default(), MainDeviceConfig::default());
+    /// # async {
+    /// # let mut group = maindevice
+    /// #     .init_single_group::<8, 8>(ethercat_now)
+    /// #     .await
+    /// #     .expect("Init");
+    /// let subdevice = group.subdevice(&maindevice, 0).expect("No subdevice!");
+    ///
+    /// // Reading the TxPDO assignment
+    /// // This is equivalent to...
+    /// let values = subdevice.sdo_read_array::<u16>(0x1c13).await?;
+    ///
+    /// // ... this
+    /// // let len: u8 = subdevice.sdo_read(0x1c13, 0).await?;
+    /// // let value1: u16 = subdevice.sdo_read(0x1c13, 1).await?;
+    /// // let value2: u16 = subdevice.sdo_read(0x1c13, 2).await?;
+    /// // let value3: u16 = subdevice.sdo_read(0x1c13, 3).await?;
+    ///
+    /// # Ok::<(), ethercrab::error::Error>(())
+    /// # };
+    /// ```
+    pub async fn sdo_read_array<T>(&self, index: u16) -> Result<Vec<T>, Error>
+    where
+        T: EtherCrabWireReadSized,
+    {
+        let len: u8 = self.sdo_read(index, 0).await?;
+
+        let mut values = Vec::with_capacity(usize::from(len));
+
+        for i in 1..=len {
+            let value: T = self.sdo_read(index, i).await?;
+            values.push(value);
+        }
+
+        Ok(values)
+    }
+
+    #[cfg(not(feature = "std"))]
```

Comment:
I'd like to see this split removed, at least for now. It adds extra API surface and documentation that needs maintaining and testing. It's not _too_ bad to convert from `heapless::Vec` into a `std::Vec` if you need to so I think it's ok to only have one variant of `sdo_read_array`. Let me know what you think.

```suggestion
```

Reply:
Yes I think having a slim codebase is nice. On the other side it's unfortunate to have an extra surface where errors can originate from. Luckily the arrays are relatively short and predictable in size so it's not that bad in this case so we can leave this out.

Reply:
Because the length is a `u8`, you could do `sdo_read_array::<_, u8::MAX as usize>::(...).await.unwrap()`. The compiler should hopefully be smart enough to optimise out the panic.